p2p/nodestate/nodestate.go

```diff
@@ -802,12 +869,26 @@ func (ns *NodeStateMachine) SetField(n *enode.Node, field Field, value interface
 	}
 
 	state := node.state
-	ns.lock.Unlock()
-	if len(f.subs) > 0 {
-		for _, cb := range f.subs {
-			cb(n, Flags{mask: state, setup: ns.setup}, oldValue, value)
+	callback := func() {
+		if len(f.subs) > 0 {
+			for _, cb := range f.subs {
+				cb(n, Flags{mask: state, setup: ns.setup}, oldValue, value, caller)
+			}
 		}
 	}
+	callNow := caller == nil
+	if callNow {
+		caller = &Caller{}
+	} else {
+		caller.pending = append(caller.pending, callback)
+		ns.callbackCount++
+	}
+	ns.lock.Unlock()
+	// call state update subscription callbacks without holding the mutex
+	if callNow {
+		callback()
+		ns.processCallbacks(caller)
+	}
```

Comment:
Isn't this equivalent? 
```suggestion
	if caller != nil {
		caller.pending = append(caller.pending, callback)
		ns.callbackCount++
	} 
	ns.lock.Unlock()
	// call state update subscription callbacks without holding the mutex
	if caller != nil {
		callback()
 		ns.processCallbacks(caller)
	}
```

Reply:
No, they are not equivalent (the `caller = &Caller{}` part is entirely missing) but this part has now changed anyways.